network/p2p/capabilities.go

```diff
@@ -0,0 +1,170 @@
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package p2p
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	"github.com/libp2p/go-libp2p/core/discovery"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	libpeerstore "github.com/libp2p/go-libp2p/core/peerstore"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/logging"
+	algoDht "github.com/algorand/go-algorand/network/p2p/dht"
+	"github.com/algorand/go-algorand/network/p2p/peerstore"
+)
+
+// Capability represents functions that some nodes may provide and other nodes would want to know about
+type Capability string
+
+const (
+	// Archival nodes
+	Archival Capability = "archival"
+	// Catchpoints storing nodes
+	Catchpoints = "catchpointStoring"
+)
+
+const operationTimeout = time.Second * 5
+const advertisementInterval = time.Hour * 22
+
+// CapabilitiesDiscovery exposes Discovery interfaces and wraps underlying DHT methods to provide capabilities advertisement for the node
+type CapabilitiesDiscovery struct {
+	disc discovery.Discovery
+	dht  *dht.IpfsDHT
+	log  logging.Logger
+	wg   sync.WaitGroup
+}
+
+// Advertise implements the discovery.Discovery/discovery.Advertiser interface
+func (c *CapabilitiesDiscovery) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	return c.disc.Advertise(ctx, ns, opts...)
+}
+
+// FindPeers implements the discovery.Discovery/discovery.Discoverer interface
+func (c *CapabilitiesDiscovery) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	return c.disc.FindPeers(ctx, ns, opts...)
+}
+
+// Close should be called when fully shutting down the node
+func (c *CapabilitiesDiscovery) Close() {
+	_ = c.dht.Close()
+	c.wg.Wait()
+}
+
+// Host exposes the underlying libp2p host.Host object
+func (c *CapabilitiesDiscovery) Host() host.Host {
+	return c.dht.Host()
+}
+
+// AddPeer adds a given peer.AddrInfo to the Host's Peerstore, and the DHT's routing table
+func (c *CapabilitiesDiscovery) AddPeer(p peer.AddrInfo) (bool, error) {
+	c.Host().Peerstore().AddAddrs(p.ID, p.Addrs, libpeerstore.TempAddrTTL)
+	return c.dht.RoutingTable().TryAddPeer(p.ID, true, true)
+}
+
+// PeersForCapability returns a slice of peer.AddrInfo for a Capability
+// Since CapabilitiesDiscovery uses a backoffcache, it will attempt to hit cache, then disk, then network
+// in order to fetch n peers which are advertising the required capability.
+func (c *CapabilitiesDiscovery) PeersForCapability(capability Capability, n int) ([]peer.AddrInfo, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), operationTimeout)
+	defer cancel()
+	var peers []peer.AddrInfo
+	peersChan, err := c.FindPeers(ctx, string(capability), discovery.Limit(n))
+	if err != nil {
+		return nil, err
+	}
+	for p := range peersChan {
+		if p.ID.Size() > 0 {
+			peers = append(peers, p)
```

Comment:
do we need to check that p.ID != node.peerID? does FindPeers return addrInfo for the host node?

Reply:
I had assumed this would be done for us, but I've gone back and looked through the code and I can't find anywhere that they would filter that out. I'll add a test to see what happens and remove the listen address that the Host is listening on from the returned list if it ends up being possible since I think you're right--we don't really ever want to get ourselves in this function call.

Reply:
Done--the new test indeed found itself before I updated this.